dev-docs/design/how-lit-html-works.md

```diff
@@ -0,0 +1,431 @@
+
+# Foundational Pieces
+
+lit-html's design arises naturally out of combining two powerful web primitives: [JavaScript Tagged Template Literals](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Template_literals#tagged_templates) (TTLs) and the HTML [`<template>`](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/template) element.
+
+TTLs provide an explicit syntax for separating static strings of HTML content and dynamic JavaScript values. The static strings are immutable and preserve identity across multiple evaluations of the tagged literal making them ideal for use as a cache key.
+
+The `<template>` element provides a container for HTML content that can later be cloned into the document. In a template element: scripts don't run, styles don't apply, custom elements don't upgrade, etc. During rendering, a template's content can be cloned or imported into the Document which makes the cloned nodes active.
+
+# Rendering
+
+## Summary of lit-html rendering phases
+
+lit-html templates are only a description of the UI. They must be rendered with `render()` to affect the DOM. The **Define** and **Render** phases are controlled by the user of Lit, while **Prepare**, **Create**, and **Update** are internal phases.
+
+<ol>
+<li>
+
+**Define**: `` const ui = (title) => html`<h1>${title}</h1>` ``
+
+</li>
+<li>
+
+**Render**: `render(ui("Example Title"), document.body)`
+
+</li>
+<ol type="a">
+<li>
+
+**Prepare**: The first time render is called for each unique TTL, create a `<template>` element containing the static HTML and Lit `TemplatePart`s encoding a mapping for each dynamic JavaScript value to its location in the static HTML.
+
+</li>
+<li>
+
+**Create**: The first time render is called for a given node and template, clone the `<template>` contents into the DOM and instantiate the Lit `Part`s which manage the dynamic content.
+
+</li>
+
+<li>
+
+**Update**: Iterate over the dynamic JS values and associated Lit `Part` only committing the values that have changed to the DOM.
+
+</li>
+</ol>
+</ol>
+
+The end result is `<h1>Example Title</h1>` rendered in the `<body>` element. If the user now executes `render(ui("Updated"), document.body)`, only the **Update** phase runs skipping the **Prepare** and **Create** phases.
+
+## Motivating example
+
+This example will be referred to throughout the rest of the document to concretely show what lit-html is doing during each internal rendering phase.
+
+```html
+<!-- File: index.html -->
+<script type="module" src="./example.js"></script>
+<style>
+  .odd {
+    color: green;
+    text-decoration: underline;
+  }
+</style>
+
+<div id="container"></div>
+```
+
+```js
+// File: example.js
+import {html, render} from 'lit';
+
+const container = document.querySelector('#container');
+
+const counterUi = (count) => html` <span class="${count % 2 == 1 ? 'odd' : ''}">
```

Comment:
An example that reuses the same template in multiple places might more clearly express the relationship between things. Consider these mermaid charts:

For the initial render:
```mermaid
flowchart TD
    A[&#40url&#41 =&gt html`&ltimg src=$&#123url&#125&gt`] -->|Called with '1.jpg'| C1(TemplateResult\n&#123strings, values: &#91'1.jpg'&#93&#125)
    A -->|Called with '2.jpg'| C2(TemplateResult\n&#123strings, values: &#91'2.jpg'&#93&#125)
    A -->|Called with '3.jpg'| C3(TemplateResult\n&#123strings, values: &#91'3.jpg'&#93&#125)
    C1 --> |Prepare| Template(&lttemplate&gt)
    C2 --> |Cache lookup| Template
    C3 --> |Cache lookup| Template
    Template --> |Create| TI1(Template Instance 1)
    Template --> |Create| TI2(Template Instance 2)  
    Template --> |Create| TI3(Template Instance 3)
    TI1 --> |Update| Up1(Render &ltimg src='1.jpg'&gt)
    TI2 --> |Update| Up2(Render &ltimg src='2.jpg'&gt)
    TI3 --> |Update| Up3(Render &ltimg src='3.jpg'&gt)  
```


For an update:
```mermaid
flowchart TD
    A[&#40url&#41 =&gt html`&ltimg src=$&#123url&#125&gt`] -->|Called with '4.jpg'| C1(TemplateResult\n&#123strings, values: &#91'4.jpg'&#93&#125)
    A -->|Called with '5.jpg'| C2(TemplateResult\n&#123strings, values: &#91'5.jpg'&#93&#125)
    A -->|Called with '6.jpg'| C3(TemplateResult\n&#123strings, values: &#91'6.jpg'&#93&#125)
    C1 --> |Cache lookup| TI1(Template Instance 1)
    C2 --> |Cache lookup| TI2(Template Instance 2)  
    C3 --> |Cache lookup| TI3(Template Instance 3)
    TI1 --> |Update| Up1(setAttribute&#40imgElem1, '4.jpg'&#41)
    TI2 --> |Update| Up2(setAttribute&#40imgElem1, '5.jpg'&#41)
    TI3 --> |Update| Up3(setAttribute&#40imgElem1, '6.jpg'&#41)
```

Reply:
I'll add this as an appendix entry. I like the current example because it demonstrates a couple different Parts.
I do like the idea of adding a mermaid chart up top to summarize the whole rendering flow.

Reply:
Added to appendix! You are the best!